src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/QueryProjectPropertiesContext.cs

```diff
@@ -0,0 +1,102 @@
+using Microsoft.VisualStudio.ProjectSystem.Query;
+using Microsoft.VisualStudio.ProjectSystem.Query.ProjectModel.Implementation;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Query
+{
+    /// <summary>
+    /// An implementation of <see cref="IProjectPropertiesContext"/> for use in the
+    /// Project Query API implementation.
+    /// </summary>
+    /// <remarks>
+    /// The point here is to capture the context for a part of a query execution in a
+    /// manner that can be passed from one provider to another and is also suitable as a
+    /// key into a cache (such as the <see cref="IPropertyPageQueryCache"/>).
+    /// </remarks>
+    internal sealed class QueryProjectPropertiesContext : IProjectPropertiesContext, IEquatable<QueryProjectPropertiesContext>
+    {
+        /// <summary>
+        /// A well-known context representing the project file as a whole.
+        /// </summary>
+        /// <remarks>
+        /// Note that if an <see cref="IProjectPropertiesContext"/> has the <see cref="IProjectPropertiesContext.IsProjectFile"/>
+        /// property is set to <c>true</c> and the <see cref="IProjectPropertiesContext.ItemType"/>
+        /// and <see cref="IProjectPropertiesContext.ItemName"/> properties are <c>null</c>
+        /// then the properties system treats it as referring to the project file as a whole
+        /// regardless of the <see cref="IProjectPropertiesContext.File"/> property. This
+        /// lets us get away with setting it to the empty string and re-using the same
+        /// instance across projects.
+        /// </remarks>
+        public static readonly QueryProjectPropertiesContext ProjectFile = new(isProjectFile: true, file: string.Empty, itemType: null, itemName: null);
+
+        public QueryProjectPropertiesContext(bool isProjectFile, string file, string? itemType, string? itemName)
+        {
+            IsProjectFile = isProjectFile;
+            File = file;
+            ItemType = itemType;
+            ItemName = itemName;
+        }
+
+        public bool IsProjectFile { get; }
+
+        public string File { get; }
+
+        public string? ItemType { get; }
+
+        public string? ItemName { get; }
+
+        public bool Equals(QueryProjectPropertiesContext? other)
+        {
+            return other is not null
+                && IsProjectFile == other.IsProjectFile
+                && StringComparers.Paths.Equals(File, other.File)
+                && StringComparers.ItemTypes.Equals(ItemType, other.ItemType)
+                && StringComparers.ItemNames.Equals(ItemName, other.ItemName);
+        }
+
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as QueryProjectPropertiesContext);
+        }
+
+        public override int GetHashCode()
+        {
+            int hashCode = IsProjectFile.GetHashCode();
+            hashCode = (hashCode * -1521134295) + StringComparers.Paths.GetHashCode(File);
```

Comment:
>1521134295 [](start = 36, length = 10)

Should this implementation use System.HashCode.Combine?

Reply:
That wasn't introduced until .NET Core 2.1.

Reply:
Ah, I didn't realize that!